lib/puppet/resource_api/property.rb

```diff
@@ -99,14 +99,28 @@ def def_custom_insync?
       title       = @resource.rsapi_title
 
       raise(Puppet::DevError, 'No insync? method defined in the provider; an insync? method must be defined if the custom_insync feature is defined for the type') unless provider.respond_to?(:insync?)
+
       provider_insync_result = provider.insync?(context, title, @attribute_name, is_hash, should_hash)
-      # if nil, super and rely on Puppet::Property.insync?
-      # if true/false, return that and use the default change messaging
-      # if string, set value of custom change reporting and treat as false
-      @change_to_s_value = provider_insync_result if provider_insync_result.is_a?(String)
-      return provider_insync_result == true unless provider_insync_result.nil?
 
-      super(is)
+      case provider_insync_result
+      when nil
+        # If validating ensure and no custom insync was used, check if rs_value matches is.
+        return rs_value.to_s == is.to_s if @attribute_name == :ensure
+        # Otherwise, super and rely on Puppet::Property.insync?
+        super(is)
+      when TrueClass, FalseClass
+        # When returning an actual boolean, use default change messaging
+        return provider_insync_result
+      when String
+        # When returning a string, set the change message to the result and return false
+        @change_to_s_value = provider_insync_result
+        return false
+      else
+        # When returning anything else, warn about a non-idiomatic return, set the change message, and return false
+        context.warning("Custom insync for #{@attribute_name} returned a #{provider_insync_result.class} instead of a String")
+        @change_to_s_value = "#{@attribute_name} changed #{is_to_s(is)} to #{provider_insync_result.inspect}"
+        return false
```

Comment:
if we're already providing a `warning` here, we might want to fall back to the default message instead? Given that we consider this an invalid use of the API, it's gonna be awkward either way.

Reply:
I'm fine with that, I think it works out to be functionally the same anyway.